storage/storage_image.go

```diff
@@ -345,10 +345,10 @@ func (s *storageImageDestination) Close() error {
 }
 
 func (s *storageImageDestination) DesiredLayerCompression() types.LayerCompression {
-	// We ultimately have to decompress layers to populate trees on disk,
-	// so callers shouldn't bother compressing them before handing them to
-	// us, if they're not already compressed.
-	return types.PreserveOriginal
+	// We ultimately have to decompress layers to populate trees on disk
+	// and need to explicitly ask for it here, so that the layers' MIME
+	// types can be set accordingly.
+	return types.Decompress
```

Comment:
@mtrmac, FYI - this was also necessary to fix the (docker -> storage -> dir) cycle

Reply:
This is definitely conceptually correct, but it also significantly changes the assumptions of the code, e.g. see the `RecordDigestUncompressedPair` call in `PutBlob`, and I’m sure the `LayersByCompressedDigest` calls get the “compressed digest” value from something in this file. I don’t think this one-liner alone suffices (but I haven’t actually re-read all of the code now).

Reply:
Or maybe it would be easier to keep the current approach, but remap the MIME types in the source’s `LayerInfosForCopy`. Easier, not necessarily cleaner.

Reply:
I'll check that. It feels like there are a bunch of traps but we're slowly getting there. I will add tests to Skopeo to make sure we're not regressing etc.

Reply:
Also, IIRC, decompressing the layers in generic code causes `PutManifest` to be called with an updated manifest, and that in turn breaks digest references using the digest from the original registry.

It’s all one big knot of entangled wool…

One way out of that could be to just not validate the digests (see also https://github.com/cri-o/cri-o/issues/2157 ); I also have an old branch that differentiates `{Put,Get}{,Original}Manifest` so that c/storage could record _both_ the original manifest (with the original digest, and which can be verified using signatures) and a modified manifest that actually corresponds to the uncompressed layers lying somewhere. In the meantime the `LayerInfosForCopy` hack emerged instead, and I’m pretty torn WRT continuing to rely on it more and more vs. the significant rework for the two-manifests approach.

Reply:
(I don’t think this has been solved yet.)

Reply:
I think we can keep lying :^) The (docker -> storage -> dir) cycle works without the change. Am I missing some potential side effect?

Reply:
AFAICS `LayerInfosForCopy` does return the uncompressed MIME type, so keeping lying should be transparent to both the current compress/decompress decision logic and the new MIME type editing logic.